File: ScratchServer.rb
#! /usr/bin/ruby
# coding: utf-8

require 'rubygems'
require 'serialport'
require 'socket'
require 'json'

port =  "/dev/ttyACM0"
sp = SerialPort.new(port, 9600, 8, 1, SerialPort::NONE)

server = TCPServer.open(51234)

print("Start Mugbot Scratch Server...\n")

while true
  Thread.start(server.accept) do | sock |
    system(`/home/pi/mugbot-talk-1.1.sh スクラッチとの接続を開始しました`)
    # for speak in English
    # system(`espeak -ven+f3 -k5 -s150 "Scratch connection established"`)
    while buf = sock.gets
      begin
        json = JSON.parse(buf)
        sock.write(buf)
        case json['action']
        when 'face_y' then
          y = json['arg'].to_i + 95
          if y > 110 then
            y = 110
          elsif y < 80 then
            y = 80
          end
          sp.write(y.to_s + "y")
          sleep(0.01)
        when 'face_x' then
          x = json['arg'].to_i + 90
          if x > 175 then
            x = 175
          elsif x < 5 then
            x = 5 
          end
          sp.write(x.to_s + "x")
          sleep(0.01)
        when 'eye' then
          z = json['arg'].to_i;
          if z > 255 then
            z = 255
          elsif x < 0 then
            x = 0
          end
          sp.write(z.to_s + "z")
          sleep(0.01)
        when 'speech' then
          sp.putc "t"
          system(`/home/pi/mugbot-talk-1.1.sh #{json['arg'].to_s}`)
          # for speak in English
          # system(`espeak -ven+f3 -k5 -s150 "#{json['arg'].to_s}"`)
          sp.putc "k"
        else
          print("Error\n")
        end
      rescue Exception => e
        $stderr.puts e
      end
    end
    sock.close
  end
end

server.close

File: ScratchServer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

try:
    import socketserver
except:
    import SocketServer as socketserver
import signal
import socket
import serial
import os
import json
import sys

HOST, PORT = '0.0.0.0', 51234
serial = serial.Serial('/dev/ttyACM0',  57600)

class ScratchHandler(socketserver.BaseRequestHandler):
    def setup(self):
        os.system('/home/pi/mugbot-talk-1.1.sh ' + 'スクラッチとの接続を開始しました &')
        # for speak in English
        # os.system('espeak -ven+f3 -k5 -s150 "Scratch connection established" &')
        
    def handle(self):
        while True:
            self.data = self.request.recv(1024).strip()
            if len(self.data) == 0:
                break
            json_obj = json.loads(self.data)
            action = json_obj['action']
            arg = json_obj['arg']
            if action == 'face_y':
                arg = min(max(int(arg) + 95, 80), 110)
                serial.write((str(arg) + 'y').encode())
            elif action == 'face_x':
                arg = min(max(int(arg) + 90, 5), 175)
                serial.write((str(arg) + 'x').encode())
            elif action == 'eye':
                arg = min(max(int(arg), 0), 255)
                serial.write((str(arg) + 'z').encode())
            elif action == 'speech':
                serial.write('t'.encode())
                if sys.version_info.major == 2:
                    arg = arg.encode('utf-8')
                os.system('/home/pi/mugbot-talk-1.1.sh ' + arg + ' &')
                # for speak in English
                # os.system('espeak -ven+f3 -k5 -s150 ' + '"' + arg +'" &')
                serial.write('n'.encode())
            else:
                print('Unknown Command')
            
class ScratchServer(socketserver.ThreadingTCPServer):
    def server_bind(self):
        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
        self.socket.bind(self.server_address)
        
if __name__ == '__main__':
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    server = ScratchServer((HOST, PORT), ScratchHandler)
    server.serve_forever()
